File: 2024/day06.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <string>
#include <cstdint>
#include <cstring>
#include <vector>
#include <map>
#include <set>
#include <queue>
#include <algorithm>
#include <numeric>

#include "utils.h"

using namespace std;

const string test(
"....#.....\n"
".........#\n"
"..........\n"
"..#.......\n"
".......#..\n"
"..........\n"
".#..^.....\n"
"........#.\n"
"#.........\n"
"......#..."
);

bool DEBUG = false;
bool TEST = false;

int WIDTH = -1;
int HEIGHT = -1;

typedef Point<short> point_t;

point_t NORTH(0,-1);
point_t EAST(-1,0);
point_t SOUTH(0,1);
point_t WEST(1,0);

point_t GUARD(-1,-1);

inline point_t turn_right( const point_t in )
{
    return point_t( -in.y, in.x );
}

struct Cell {
    int walkId;
    bool blocked;
    map<point_t, bool> walked;

    Cell()
        : walkId( -1 )
        , blocked( false )
    {
        maybeResetCell( 0 );
    }

    void maybeResetCell(int newwalk)
    {
        if( walkId != newwalk )
        {
            walkId = newwalk;
            walked[NORTH] = false;
            walked[EAST] = false;
            walked[SOUTH] = false;
            walked[WEST] = false;
        }
    }
};

typedef vector<vector<Cell>> grid_t;

void parseInput( const StringVector & data, grid_t & grid )
{
    grid.resize(HEIGHT);

    for( int y=0; y < HEIGHT; y++ )
    {
        grid[y].resize(WIDTH);
        for( int x=0; x < WIDTH; x++ )
        {
            char c = data[y][x];
            if( c == '#' )
                grid[y][x].blocked = true;
            else if( c == '^' )
                GUARD = point_t(x,y);
        }
    }
}

// A "path" step includes both a coordinate and a direction.

struct step_t
{
    point_t pt;
    point_t dir;
    bool operator==( step_t & other )
    {
        return pt==other.pt && dir==other.dir;
    }
    bool operator!=( step_t & other )
    {
        return !(*this==other);
    }
};

// Ugliness warning -- this gets passed from pass 1 to pass 2.

vector<step_t> g_originalPath;

int part1( StringVector & data, grid_t & grid )
{
    point_t gpt = GUARD;
    point_t dir = NORTH;
    g_originalPath.clear();

    for(;;)
    {
        g_originalPath.push_back({gpt,dir});
        point_t npt = gpt + dir;
        if( !(between(0, npt.x, WIDTH) && between(0, npt.y, HEIGHT)) )
            break;
        if( grid[npt.y][npt.x].blocked )
            dir = turn_right(dir);
        else
            gpt = npt;
    }

    // Count the unique cells in the path.

    set<point_t> visited;
    transform(
        g_originalPath.begin(),
        g_originalPath.end(),
        inserter(visited, visited.begin()),
        [](step_t & pth) { return pth.pt; }
    );

    return visited.size();
}


int walkCandidateMap( grid_t & grid,  step_t & point )
{
    static int currentWalk = 0;
    currentWalk++;
    point_t pt = point.pt;
    point_t direction = point.dir;

    for(;;)
    {
        // We keep walking until we get blocked.
        point_t npt = pt + direction;
        if( !(between(0, npt.x, WIDTH) && between(0, npt.y, HEIGHT)) )
            break;
        if( !grid[npt.y][npt.x].blocked )
        {
            pt = npt;
            continue;
        }
        Cell & currentCell = grid[pt.y][pt.x];
        currentCell.maybeResetCell(currentWalk);
        if( currentCell.walked[direction] )
            return 1;
        currentCell.walked[direction] = true;
        direction = turn_right(direction);
    }
    return 0;
}


int part2( StringVector & data, grid_t & grid )
{
    set<point_t> checked;
    checked.insert( GUARD );
    int countOfLoopPaths = 0;
    step_t point = g_originalPath[0];
    for( auto & step : g_originalPath )
    {
        if( step != point )
        {
            point_t block = step.pt;
            if( checked.find(block) == checked.end() )
            {
                checked.insert( block );
                grid[block.y][block.x].blocked = true;
                countOfLoopPaths += walkCandidateMap(grid, point);
                grid[block.y][block.x].blocked = false;
            }
        }
        point = step;
    }

    return countOfLoopPaths;
}


int main( int argc, char ** argv )
{
    while( *++argv )
    {
        string arg(*argv);
        if( arg == "debug")
            DEBUG = true;
        else if( arg =="test")
            TEST = true;
    }

    string input = TEST ? test : file_contents("day06.txt");
    
    StringVector data = split( input, "\n");
    WIDTH = data[0].size();
    HEIGHT = data.size();

    vector<vector<Cell>> grid;
    parseInput( data, grid );

    cout << "Part 1: " << part1(data,grid) << "\n";
    cout << "Part 2: " << part2(data,grid) << "\n";
}

File: 2024/day06.go
package main

import (
	"fmt"
	"strings"

	"aoc/tools"
	_ "embed"
)

var DEBUG bool = false
var TEST bool = false

var test = `
....#.....
.........#
..........
..#.......
.......#..
..........
.#..^.....
........#.
#.........
......#...`[1:]

//go:embed day06.txt
var live string

var WIDTH int = -1
var HEIGHT int = -1

type Point = tools.Point

var NORTH Point = Point{0, -1}
var EAST Point = Point{-1, 0}
var SOUTH Point = Point{0, 1}
var WEST Point = Point{1, 0}

var GUARD Point = Point{-1, -1}

type Cell struct {
	walkId  int
	blocked bool
	walked  map[Point]bool
}

func (c *Cell) maybeResetCell(newwalk int) {
	if c.walkId != newwalk {
		c.walkId = newwalk
		c.walked[NORTH] = false
		c.walked[EAST] = false
		c.walked[SOUTH] = false
		c.walked[WEST] = false
	}
}

func makeCell() Cell {
	c := Cell{-1, false, make(map[Point]bool, 0)}
	c.maybeResetCell(0)
	return c
}

func parseInput(data []string) [][]Cell {
	WIDTH = len(data[0])
	HEIGHT = len(data)

	grid := make([][]Cell, HEIGHT)
	for y := 0; y < HEIGHT; y++ {
		grid[y] = make([]Cell, WIDTH)
		for x := 0; x < WIDTH; x++ {
			grid[y][x] = makeCell()
			var c = data[y][x]
			if c == '#' {
				grid[y][x].blocked = true
			} else if c == '^' {
				GUARD = Point{x, y}
			}
		}
	}
	return grid
}

// A path step includes both a coordinate and a direction.

type Step struct {
	pt  Point
	dir Point
}

// Ugliness warning -- this gets passed from pass 1 to pass 2.

var g_originalPath []Step

func part1(grid [][]Cell) int {
	gpt := GUARD
	dir := NORTH

	g_originalPath = make([]Step, 0)

	for {
		g_originalPath = append(g_originalPath, Step{gpt, dir})
		npt := gpt.Add(dir)
		if !npt.InRange(WIDTH, HEIGHT) {
			break
		}
		if grid[npt.Y][npt.X].blocked {
			dir = dir.Right()
		} else {
			gpt = npt
		}
	}

	// Count the unique cells in the path.

	visited := make(map[Point]bool)
	for _, pth := range g_originalPath {
		visited[pth.pt] = true
	}

	return len(visited)
}

var currentWalk int = 0

func walkCandidateMap(grid [][]Cell, point Step) int {
	currentWalk++
	pt := point.pt
	direction := point.dir

	for {
		// We keep walking until we get blocked.
		npt := pt.Add(direction)
		if !npt.InRange(WIDTH, HEIGHT) {
			break
		}
		if !grid[npt.Y][npt.X].blocked {
			pt = npt
			continue
		}
		currentCell := &grid[pt.Y][pt.X]
		currentCell.maybeResetCell(currentWalk)
		if currentCell.walked[direction] {
			return 1
		}
		currentCell.walked[direction] = true
		direction = direction.Right()
	}
	return 0
}

func part2(grid [][]Cell) int {
	checked := make(map[Point]bool)
	checked[GUARD] = true
	countOfLoopPaths := 0
	point := g_originalPath[0]
	for _, step := range g_originalPath {
		if step != point {
			block := step.pt
			if !checked[block] {
				checked[block] = true
				grid[block.Y][block.X].blocked = true
				countOfLoopPaths += walkCandidateMap(grid, point)
				grid[block.Y][block.X].blocked = false
			}
		}
		point = step
	}

	return countOfLoopPaths
}

func main() {
	var input string
	TEST, DEBUG, input = tools.Setup(test, live)
	grid := parseInput(strings.Split(input, "\n"))

	fmt.Println("Part 1:", part1(grid))
	fmt.Println("Part 2:", part2(grid))
}
